File: leetcode/c++/43. Multiply Strings.cpp
#include<iostream>
#include <bits/stdc++.h>

using namespace std;

string multiply(string nums1,string nums2){
  reverse(nums1.begin(),nums1.end());
  reverse(nums2.begin(),nums2.end());
  vector<int> d(nums1.length()+nums2.length());
  for (int i = 0; i < nums1.length(); i++) {
    for (int j = 0; j < nums2.length(); j++) {
      d[i + j] += (nums1[i] - '0') * (nums2[j] - '0');
    }
  }
  string result = "";
  for (int i = 0; i < d.size(); i++) {
    int digit = d[i] % 10;
    int carry = d[i] / 10;
    if (i + 1 < d.size()) {
        d[i + 1] += carry;
    }
    result = to_string(digit) + result;
  }
  int index=0;
  while (index < result.length() - 1 && result[index] == '0'){
    index++;
  }
  return result.substr(index);
}

int main(){
  cout << multiply("7","8")<< endl;
  return 0;
}

File: leetcode/kotlin/43. Multiply Strings.kt
import java.io.*

internal object Main {
    var ps: PrintStream = System.out
    fun multiply(num1: String?, num2: String?): String {
        val n1: String = StringBuilder(num1).reverse().toString()
        val n2: String = StringBuilder(num2).reverse().toString()
        val d = IntArray(n1.length() + n2.length())
        for (i in 0 until n1.length()) {
            for (j in 0 until n2.length()) {
                d[i + j] += (n1.charAt(i) - '0') * (n2.charAt(j) - '0')
            }
        }
        var result = ""
        for (i in d.indices) {
            val digit = d[i] % 10
            val carry = d[i] / 10
            if (i + 1 < d.size) {
                d[i + 1] += carry
            }
            result = digit.toString() + "" + result
        }
        var index = 0
        while (index < result.length() - 1 && result.charAt(index) === '0') {
            index++
        }
        return result.substring(index)
    }

    fun main(vararg s: String?) {
        ps.println(multiply("7", "8"))
    }
}